### PR TITLE
[visualization] Add render camera preview

### PR DESCRIPTION
**Pull request by jwnimmer-tri**
Caveat: This won't work in Deepnote until #18862 and related are finished.  It uses a GL native window as the preview for now.

[Screencast from 04-07-2023 12:48:36 PM.webm](https://user-images.githubusercontent.com/17596505/230668625-da9f88df-2c4b-4b2b-9e62-025b3c5b9897.webm)

- - -
This change is [Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19158)

### CODE REVIEW

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19158)* status: 1 unresolved discussion, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @jwnimmer-tri)

___
*[`bindings/pydrake/visualization/_rgbd_camera_sliders.sdf` line 1 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/19158#-NSfuHTE0LtYnUn5xZi_:-NSfuHTE0LtYnUn5xZia:b-gzgs97) ([raw file](https://github.com/RobotLocomotion/drake/blob/3e9749b95886719ebffe3577363b82138f23b491/bindings/pydrake/visualization/_rgbd_camera_sliders.sdf#L1)):*
> ```
> <?xml version="1.0"?>
> ```

Working

Installing a model file defined inside of `bindings/...` is non-trivial.  I need to rework this somehow.

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19158)* status: 1 unresolved discussion, needs platform reviewer assigned, needs at least two assigned reviewers (waiting on @jwnimmer-tri)

___
*[`bindings/pydrake/visualization/_rgbd_camera_sliders.sdf` line 80 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/19158#-NSl9k-fA_X5G2SX4quk:-NSl9k-g5xu_JEvG1WLk:bnyoh6a) ([raw file](https://github.com/RobotLocomotion/drake/blob/eb6fb0a7e0365ebcd8bf6e9f1cbeb0574ff5b735/bindings/pydrake/visualization/_rgbd_camera_sliders.sdf#L80)):*
> ```
>       </axis>
>     </joint>
>     <joint name="$rgbd_zenith" type="revolute">
> ```

Working

I'm calling this "zenith" but it's actually "90 - zenith", i.e., with the opposite sign.  That _seemed_ the most natural compromise for my own UX taste (so that positive slider movement moves the camera up in elevation), but I wonder if it might confuse some very-precise readers that it's not an actual zenith.  Possibly I should rename it, or switch its sign here.

Any votes?

**Review by SeanCurtis-TRI — COMMENTED** [review]
:LGTM: With some cosmetic thoughts.

> Reviewed 2 of 5 files at r1, 6 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19158#-:-NTGIRqz7GGPQBvT_4rf:b-gi9p6k)* status: 3 unresolved discussions, needs at least two assigned reviewers (waiting on @jwnimmer-tri)

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/19158#-NTGHsB_EBEXsB0UpH3o:-NTGHsBa7BJ30piEbOLW:by0d73t) (no related file):*
With this added code, I seem to get more instances of the following lines in my console:


```
LCM self test failed!!
Check your routing tables and firewall settings
```

With this PR, it happens consistently twice. Prior to this, at most once.

___
___
*[`bindings/pydrake/visualization/_rgbd_camera_sliders.sdf` line 36 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/19158#-NTGFZoW9arkiRCSswCe:-NTGFZoW9arkiRCSswCf:b-91m4rf) ([raw file](https://github.com/RobotLocomotion/drake/blob/d4b1d7615b508ef05f13bd4b744b73641bbdb3d7/bindings/pydrake/visualization/_rgbd_camera_sliders.sdf#L36)):*
> ```
>       <child>base</child>
>     </joint>
>     <joint name="$rgbd_target_x" type="prismatic">
> ```

BTW This will be really subtle, but I'll try.

tl;dr feels like "target" is the incorrect name.  "pivot" or "origin" would be possibly be better.

In most APIs when a camera is posed via a "target" parameter, the semantics is typically one of "here's a target point", "here's a camera point", "here's a camera up vector". They are nine (mostly) independent dofs from which camera extrinsics are computed. In those paradigms, moving the target does *not* change camera position (only orientation).
___
*[`bindings/pydrake/visualization/_rgbd_camera_sliders.sdf` line 80 at r2](https://reviewable.io/reviews/RobotLocomotion/drake/19158#-NSl9k-fA_X5G2SX4quk:-NTGGjGD4Dn99PN_MRcT:byqhsdr) ([raw file](https://github.com/RobotLocomotion/drake/blob/d4b1d7615b508ef05f13bd4b744b73641bbdb3d7/bindings/pydrake/visualization/_rgbd_camera_sliders.sdf#L80)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Working

I'm calling this "zenith" but it's actually "90 - zenith", i.e., with the opposite sign.  That _seemed_ the most natural compromise for my own UX taste (so that positive slider movement moves the camera up in elevation), but I wonder if it might confuse some very-precise readers that it's not an actual zenith.  Possibly I should rename it, or switch its sign here.

Any votes?
</blockquote></details>

You could go for "heading" and "tilt".

Tilt directly matches the current semantics of the slider (and is a [very "camera-y" word](https://en.wikipedia.org/wiki/Tilt_(camera))).

(Note: I avoided suggesting pan, dolly, and crane for the "target_*" values because they are quantities expressed in different frames -- world vs camera.)

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19158)* status: 1 unresolved discussion, needs at least two assigned reviewers (waiting on @jwnimmer-tri and @SeanCurtis-TRI)

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/19158#-NTGHsB_EBEXsB0UpH3o:-NTJbNRVC2-A1z0Fd9XK:b7rb69t) (no related file):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

With this added code, I seem to get more instances of the following lines in my console:


```
LCM self test failed!!
Check your routing tables and firewall settings
```

With this PR, it happens consistently twice. Prior to this, at most once.

</blockquote></details>

Working

Yeah, if you don't have your kernel configured for sending 10+ Hz image traffic over LCM, then you'll see that.  For now, my plan will be to not transmit LCM images.  (Once Meldis can show them, though, we'll probably want at least an option to transmit them.)

___

**Review by SeanCurtis-TRI — COMMENTED** [review]
Given you're on call today, if you want to grab another reviewer for feature-ish review, you can treat mine as platform.

> Reviewed 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19158#-:-NTJhz44B30wZnPHycsN:beqcf5n)* status: 1 unresolved discussion, needs at least two assigned reviewers (waiting on @jwnimmer-tri)

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/19158#-NTGHsB_EBEXsB0UpH3o:-NTJhHCA7WptABDWkjpA:b-tggofc) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Working

Yeah, if you don't have your kernel configured for sending 10+ Hz image traffic over LCM, then you'll see that.  For now, my plan will be to not transmit LCM images.  (Once Meldis can show them, though, we'll probably want at least an option to transmit them.)
</blockquote></details>

Related, there was a palpable delay between starting the process and (either) opening a new window (`-w`) or connecting to an existing window. That long window is one of the reasons why more "self test" failure messages were presented. But, it also led me to wonder if I had a bug. Only my laziness caused me to stall long enough to give it a chance to resolve (instead of hitting Ctrl-C).

What is your expectation for someone who is not quite as lazy as I am and hits Ctrl-C after four or five seconds (thinking that things are broken)?

___

**Review by jwnimmer-tri — COMMENTED** [review]
This won't be ready today anyway. I need to submit another PR first, to allow for not transmitting camera images. That one probably won't merge until tomorrow or the next day anyway.

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19158#-:-NTJjjv49FNb1xL5mCxd:b-sm4aiu)* status: 1 unresolved discussion, needs at least two assigned reviewers (waiting on @jwnimmer-tri)

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/19158#-NTGHsB_EBEXsB0UpH3o:-NTJjNQU4pyBIMQYXKz2:ba560uk) (no related file):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Related, there was a palpable delay between starting the process and (either) opening a new window (`-w`) or connecting to an existing window. That long window is one of the reasons why more "self test" failure messages were presented. But, it also led me to wonder if I had a bug. Only my laziness caused me to stall long enough to give it a chance to resolve (instead of hitting Ctrl-C).

What is your expectation for someone who is not quite as lazy as I am and hits Ctrl-C after four or five seconds (thinking that things are broken)?
</blockquote></details>

My disposition is set to "Working" and I explained my proposed change above. Once I finish my changes, I expect that the tool will no longer behave confusingly in the first place. We'll know more after I've made the change and we can re-test it.

___

**Review by SeanCurtis-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19158)* status: 1 unresolved discussion, needs at least two assigned reviewers (waiting on @jwnimmer-tri)

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/19158#-NTGHsB_EBEXsB0UpH3o:-NTOlPCA6tBirurYZPMf:b-izjdep) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The model visualizer has always offered _both_ meshcat and meldis.  Users can click the http link to open meshcat, or separately fire up meldis to see what things look like in that regime. As such, all visualization (including cameras, in the current draft here) have always gone out over LCM.  All of the `ApplyFooConfig` etc including `ApplyCameraConfig` add LCM transmitters.
</blockquote></details>

Riiiiiiight. That's it.

___

**Review by jwnimmer-tri — COMMENTED** [review]
+(status: do not review) while we want for the framework PR to land first.  You're welcome to re-test the performance / responsiveness.

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19158#-:-NTZSx5kCd-VkSrLmKxK:b-sertk3)* status: 2 unresolved discussions, needs at least two assigned reviewers, commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @jwnimmer-tri and @SeanCurtis-TRI)

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/19158#-NTGHsB_EBEXsB0UpH3o:-NTZTAEP3iY5Ebd4iJRv:b-896fix) (no related file):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Riiiiiiight. That's it.
</blockquote></details>

Done.

___
*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/19158#-NTZT75c1m-NKcTSNYsB:-NTZT75c1m-NKcTSNYsC:b-pq2aei) (no related file):*
Working

The first commit is destined to land in another PR first.

___

**Review by jwnimmer-tri — COMMENTED** [review]
-(status: do not review) 

I _think_ the LCM traffic should be back under control now.  It should only be Meldis traffic now, no images.

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19158#-:-NTp9QXq-Y6RcT30ZAAB:b-bre8r)* status: 1 unresolved discussion, needs at least two assigned reviewers (waiting on @SeanCurtis-TRI)

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/19158#-NTZT75c1m-NKcTSNYsB:-NTp9Nh2DLQZBgaXEurO:b-896fix) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Working

The first commit is destined to land in another PR first.
</blockquote></details>

Done.

___

**Review by jwnimmer-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19158)* status: 1 unresolved discussion, needs at least two assigned reviewers (waiting on @jwnimmer-tri)

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/19158#-NTGHsB_EBEXsB0UpH3o:-NV6hzN7B2NdbOpfE8Jz:b-1ifunt) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Re-ping.

___

**Comment by jwnimmer-tri** [comment]
+@xuchenhan-tri for platform review per schedule, please.

**Comment by jwnimmer-tri** [comment]
From memory, it was this:

```
bazel run //tools:model_visualizer -- --show_rgbd_sensor \
  package://drake/manipulation/models/iiwa_description/iiwa7/iiwa7_with_box_collision.sdf
```

I can double-check once I'm back at a working build tree.

**Review by xuchenhan-tri — COMMENTED** [review]
:lgtm_strong: 

> Reviewed 1 of 5 files at r1, 3 of 6 files at r2, 3 of 3 files at r3, 1 of 7 files at r4, 6 of 6 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19158#-:-NVegVaa66bJlpDaHflU:b-uwkowd)* status: 1 unresolved discussion

**Review by jwnimmer-tri — COMMENTED** [review]
> Dismissed @SeanCurtis-TRI from a discussion.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19158)* status: :shipit: complete! all discussions resolved, LGTM from assignees SeanCurtis-TRI(platform),xuchenhan-tri(platform)

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/19158#-NTGHsB_EBEXsB0UpH3o:-NVelFJE3GIRnn6L32dH:byyesvv) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Re-ping.
</blockquote></details>

I'm going to dismiss this.  I'm pretty convinced that the network overloading fix was sufficient.

___

**Review by xuchenhan-tri — COMMENTED** [review]
> Reviewed 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/19158)* status: :shipit: complete! all discussions resolved, LGTM from assignees SeanCurtis-TRI(platform),xuchenhan-tri(platform)
